File: app/src/main/java/io/stamethyst/RendererBackend.java
package io.stamethyst;

public enum RendererBackend {
    OPENGL_ES2("opengles2", "OpenGL ES2 (default)", "OpenGL ES2"),
    ANGLE("opengles2_angle", "ANGLE (OpenGL ES2)", "ANGLE"),
    MOBILEGLUES("opengles3_mobileglues", "MobileGlues (OpenGL ES3)", "MobileGlues"),
    KOPPER_ZINK("opengles3_desktopgl_zink_kopper", "Mesa Zink (Kopper)", "Mesa Zink (Kopper)");

    private final String rendererId;
    private final String selectorLabel;
    private final String statusLabel;

    RendererBackend(String rendererId, String selectorLabel, String statusLabel) {
        this.rendererId = rendererId;
        this.selectorLabel = selectorLabel;
        this.statusLabel = statusLabel;
    }

    public String rendererId() {
        return rendererId;
    }

    public String selectorLabel() {
        return selectorLabel;
    }

    public String statusLabel() {
        return statusLabel;
    }

    public String lwjglOpenGlLibName() {
        switch (this) {
            case KOPPER_ZINK:
                return "libglxshim.so";
            case ANGLE:
                return "libGLESv2_angle.so";
            case MOBILEGLUES:
                return "libmobileglues.so";
            case OPENGL_ES2:
            default:
                return "libGLESv2.so";
        }
    }

    public static RendererBackend fromRendererId(String value) {
        if (value != null) {
            String normalized = value.trim();
            for (RendererBackend backend : values()) {
                if (backend.rendererId.equalsIgnoreCase(normalized)) {
                    return backend;
                }
            }
        }
        return OPENGL_ES2;
    }
}

File: app/src/main/java/io/stamethyst/backend/render/RendererBackend.kt
package io.stamethyst.backend.render

import io.stamethyst.config.RenderSurfaceBackend

enum class RendererBackend(
    private val rendererIdValue: String,
    val displayName: String,
    private val lwjglOpenGlLibNameValue: String,
    val requiredNativeLibraries: Set<String>,
    val requiresVulkan: Boolean = false,
    val forcedSurfaceBackend: RenderSurfaceBackend? = null
) {
    OPENGL_ES_MOBILEGLUES(
        rendererIdValue = "opengles_mobileglues",
        displayName = "MobileGlues",
        lwjglOpenGlLibNameValue = "libmobileglues.so",
        requiredNativeLibraries = setOf("libmobileglues.so")
    ),
    OPENGL_ES2_NATIVE(
        rendererIdValue = "opengles2_native",
        displayName = "OpenGL ES 2",
        lwjglOpenGlLibNameValue = "libGLESv2.so",
        requiredNativeLibraries = emptySet()
    ),
    OPENGL_ES2_GL4ES(
        rendererIdValue = "opengles2",
        displayName = "GL4ES",
        lwjglOpenGlLibNameValue = "libgl4es_114.so",
        requiredNativeLibraries = setOf("libgl4es_114.so")
    ),
    OPENGL_ES3_DESKTOPGL_ZINK_KOPPER(
        rendererIdValue = "opengles3_desktopgl_zink_kopper",
        displayName = "Kopper Zink",
        lwjglOpenGlLibNameValue = "libglxshim.so",
        requiredNativeLibraries = setOf(
            "libc++_shared.so",
            "libglxshim.so",
            "libEGL_mesa.so",
            "libglapi.so",
            "libzink_dri.so"
        ),
        requiresVulkan = true,
        forcedSurfaceBackend = RenderSurfaceBackend.TEXTURE_VIEW
    ),
    VULKAN_ZINK(
        rendererIdValue = "vulkan_zink",
        displayName = "Vulkan Zink",
        lwjglOpenGlLibNameValue = "libOSMesa.so",
        requiredNativeLibraries = setOf(
            "libc++_shared.so",
            "libOSMesa.so"
        ),
        requiresVulkan = true
    );

    fun rendererId(): String = rendererIdValue

    fun lwjglOpenGlLibName(): String = lwjglOpenGlLibNameValue

    companion object {
        private const val LEGACY_RENDERER_ID_OPENGL_ES3_LTW = "opengles3_ltw"

        @JvmStatic
        fun fromRendererId(value: String?): RendererBackend? {
            if (value.isNullOrBlank()) {
                return null
            }
            val normalizedValue = value.trim()
            return when (normalizedValue) {
                // LTW was removed. GL4ES is the closest surviving backend and
                // avoids silently dropping old installs onto MobileGlues.
                LEGACY_RENDERER_ID_OPENGL_ES3_LTW -> OPENGL_ES2_GL4ES
                else -> entries.firstOrNull { it.rendererIdValue == normalizedValue }
            }
        }
    }
}
